src/Disks/BlobStorage/DiskBlobStorage.cpp

```diff
@@ -0,0 +1,194 @@
+    max_single_read_retries(max_single_read_retries_),
+    max_single_download_retries(max_single_download_retries_),
+    thread_pool_size(thread_pool_size_),
+    objects_chunk_size_to_delete(objects_chunk_size_to_delete_) {}
+
+
+class BlobStoragePathKeeper : public RemoteFSPathKeeper
+{
+public:
+    // NOTE : chunk_limit unused
+    BlobStoragePathKeeper(size_t chunk_limit_) : RemoteFSPathKeeper(chunk_limit_) {}
+
+    void addPath(const String & path) override
+    {
+        paths.push_back(path);
+    }
+
+// TODO: maybe introduce a getter?
+// private:
+    std::vector<String> paths;
+};
+
+
+DiskBlobStorage::DiskBlobStorage(
+    const String & name_,
+    DiskPtr metadata_disk_,
+    std::shared_ptr<Azure::Storage::Blobs::BlobContainerClient> blob_container_client_,
+    SettingsPtr settings_,
+    GetDiskSettings settings_getter_) :
+    IDiskRemote(name_, "" /* TODO: shall we provide a config for this path? */, metadata_disk_, "DiskBlobStorage", settings_->thread_pool_size),
+    blob_container_client(blob_container_client_),
+    current_settings(std::move(settings_)),
+    settings_getter(settings_getter_) {}
+
+
+std::unique_ptr<ReadBufferFromFileBase> DiskBlobStorage::readFile(
+    const String & path,
+    const ReadSettings & read_settings,
+    std::optional<size_t> /*estimated_size*/) const
+{
+    auto settings = current_settings.get();
+    auto metadata = readMeta(path);
+
+    LOG_TRACE(log, "Read from file by path: {}", backQuote(metadata_disk->getPath() + path));
+
+    bool threadpool_read = read_settings.remote_fs_method == RemoteFSReadMethod::threadpool;
+
+    auto reader_impl = std::make_unique<ReadBufferFromBlobStorageGather>(
+        path, blob_container_client, metadata, settings->max_single_read_retries,
+        settings->max_single_download_retries, read_settings, threadpool_read);
+
+    if (threadpool_read)
+    {
+        auto reader = getThreadPoolReader();
+        return std::make_unique<AsynchronousReadIndirectBufferFromRemoteFS>(reader, read_settings, std::move(reader_impl));
+    }
+    else
+    {
+        auto buf = std::make_unique<ReadIndirectBufferFromRemoteFS>(std::move(reader_impl));
+        return std::make_unique<SeekAvoidingReadBuffer>(std::move(buf), current_settings.get()->min_bytes_for_seek);
+    }
+}
+
+
+std::unique_ptr<WriteBufferFromFileBase> DiskBlobStorage::writeFile(
+    const String & path,
+    size_t buf_size,
+    WriteMode mode)
+{
+    auto metadata = readOrCreateMetaForWriting(path, mode);
+    auto blob_path = path; // TODO: maybe use getRandomName() or modify the path (now it contains the tmp_* directory part)
+
+    LOG_TRACE(log, "{} to file by path: {}. Blob Storage path: {}",
+        mode == WriteMode::Rewrite ? "Write" : "Append", backQuote(metadata_disk->getPath() + path), remote_fs_root_path + blob_path);
+
+    auto buffer = std::make_unique<WriteBufferFromBlobStorage>(
+        blob_container_client,
+        metadata.remote_fs_root_path + blob_path,
```

Comment:
Do you think there is a reason to set `remote_fs_root_path` as something else than an empty string in `IDiskRemote` constructor?

Reply:
But it is not set in any other place apart from constructor, but is always used.
It equals `endpoint_url` in your case. But as I see not needed for your case because you create `Azure::Storage::Blobs::BlobContainerClient`, so `endpoint_url` is not needed anymore. So ok to leave empty string. But if it is always empty then probably no need to use `metadata.remote_fs_root_path` anywhere in this disk.